core/preconditioner/isai.cpp

```diff
@@ -127,38 +138,86 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
     // system of excess blocks
     Array<IndexType> excess_row_ptrs_full{exec, num_rows + 1};
 
-    exec->run(isai::make_generate_tri_inverse(
-        lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
-        excess_row_ptrs_full.get_data(), is_lower));
+    if (is_general) {
+        exec->run(isai::make_generate_general_inverse(
+            lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
+            excess_row_ptrs_full.get_data()));
+    } else {
+        exec->run(isai::make_generate_tri_inverse(
+            lend(to_invert), lend(inverted), excess_block_ptrs.get_data(),
+            excess_row_ptrs_full.get_data(), is_lower));
+    }
 
-    auto excess_dim =
+    auto total_excess_dim =
         exec->copy_val_to_host(excess_block_ptrs.get_const_data() + num_rows);
     // if we had long rows:
-    if (excess_dim > 0) {
-        // build the excess sparse triangular system
-        auto excess_nnz = exec->copy_val_to_host(
-            excess_row_ptrs_full.get_const_data() + num_rows);
-        auto excess_system =
-            Csr::create(exec, dim<2>(excess_dim, excess_dim), excess_nnz);
-        auto excess_rhs = Dense::create(exec, dim<2>(excess_dim, 1));
-        auto excess_solution = Dense::create(exec, dim<2>(excess_dim, 1));
-        exec->run(isai::make_generate_excess_system(
-            lend(to_invert), lend(inverted), excess_block_ptrs.get_const_data(),
-            excess_row_ptrs_full.get_const_data(), lend(excess_system),
-            lend(excess_rhs)));
-        // solve it after transposing
-        std::unique_ptr<LinOpFactory> trs_factory;
-        if (is_lower) {
-            trs_factory = UpperTrs::build().on(exec);
-        } else {
-            trs_factory = LowerTrs::build().on(exec);
+    if (total_excess_dim > 0) {
+        bool done = false;
+        size_type block = 0;
+        while (true) {
+            // build the excess sparse triangular system
+            size_type excess_dim = 0;
+            size_type excess_start = block;
+            const auto block_offset = exec->copy_val_to_host(
+                excess_block_ptrs.get_const_data() + block);
+            const auto nnz_offset = exec->copy_val_to_host(
+                excess_row_ptrs_full.get_const_data() + block);
+            while (excess_dim < excess_lim && block < num_rows) {
+                block++;
+                excess_dim = exec->copy_val_to_host(
+                                 excess_block_ptrs.get_const_data() + block) -
+                             block_offset;
+            }
+            if (excess_dim == 0) break;
```

Comment:
We do a lot of individual copies here, which might not be too efficient. I would suggest that we instead create a host-side copy of excess_block_ptrs and use std::upper_bound to find the end of the current batch of blocks. Then we could also write this in a nicer loop:
```cpp
for (size_type excess_start = 0; excess_start < total_excess_dim; excess_start = find_next_block_index_lambda(excess_start))
```